runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -193,36 +193,79 @@ protected boolean discoverAddressSpace(final UUID streamId,
     private void processCheckpoint(StreamAddressSpace streamAddressSpace, Function<ILogData, BackpointerOp> filter,
                                    NavigableSet<Long> queue) {
         List<Long> checkpointAddresses = new ArrayList<>();
-        List<ILogData> checkpointData = new ArrayList<>();
         streamAddressSpace.getAddressMap().forEach(checkpointAddresses::add);
+        Collections.reverse(checkpointAddresses);
+
+        // Checkpoint entries will be read in batches of a predefined size,
+        // the reason not to read them all in a single call is that:
+        // 1. There might be more than one checkpoint and not all of them might be required.
+        // The latest checkpoint subsumes all other checkpoints (we could be reading
+        // more data than necessary).
+        // 2. Because trims are async (between sequencer and log unit), part of these
+        // addresses (for instance for the first checkpoint) might have been already
+        // trimmed from the log, but stream maps (sequencer) still do not reflect it
+        // (causing unnecessary TrimmedExceptions--as this checkpoint is not even
+        // needed in the first place).
+        Iterable<List<Long>> batches = Iterables.partition(checkpointAddresses,
+                runtime.getParameters().getCheckpointReadBatchSize());
+
+        boolean checkpointResolved = false;
+        for (List<Long> batch : batches) {
+            if (checkpointResolved) {
+                // No need to fetch next batch of addresses, checkpoint has been resolved.
+                break;
+            }
 
-        // Read all checkpoint entries in a batch read
-        try {
-            checkpointData = readAll(checkpointAddresses);
-        } catch (TrimmedException te) {
-            if (options.ignoreTrimmed) {
-                log.warn("getStreamAddressMap[{}]: Ignoring trimmed exception for address[{}]," +
-                        " stream[{}]", this, checkpointAddresses, id);
-            } else {
-                throw te;
+            try {
+                List<ILogData> entries = readAll(batch);
+                for (ILogData data : entries) {
+                    checkpointResolved = filterCheckpointEntry(data, filter, queue);
+                    if (checkpointResolved) {
+                        break;
+                    }
+                }
+            } catch (TrimmedException te) {
+                // Read one entry at a time for the last failed batch, this way we might load
+                // the required checkpoint entries until the stop condition is fulfilled
+                // without hitting a trimmed position. If we do hit a trim it is an actual
+                // TrimmedException
+                long lastReadAddress = Address.NON_ADDRESS;
+                try {
+                    for (long address : batch) {
```

Comment:
This code looks repeated. Can we refactor this to a method with a flag for single-step or batch.